{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6f6fe642-5632-449e-91c2-0e660d4a1f30",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from lxml import etree\n",
    "from collections import OrderedDict\n",
    "from urllib.parse import quote\n",
    "import csv\n",
    "import traceback\n",
    "import re, os\n",
    "from time import sleep\n",
    "from datetime import datetime\n",
    "import json\n",
    "import pandas as pd\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0c8e169e-83ce-4c15-b1e4-3cc303c0b3d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_text(raw):\n",
    "    noise_text = ['\\n',' ']\n",
    "    new = ''.join(str(s) for s in raw if s not in noise_text)\n",
    "    return new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a536f276-1f1a-4cbf-ae7e-97d6a19e9367",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_author_info(uid):\n",
    "    url = f'https://weibo.com/ajax/profile/info?custom={uid}'\n",
    "                    \n",
    "    res_author = requests.get(url, headers=headers).text\n",
    "    author_following, author_followers, author_gender, author_location, author_IPlocation = '','','','',''\n",
    "    info_dict = json.loads(res_author)['data']['user']\n",
    "    author_following = info_dict['friends_count']\n",
    "    author_followers = info_dict['followers_count']\n",
    "    author_gender = info_dict['gender']\n",
    "    author_location = info_dict['location']\n",
    "\n",
    "    return author_following, author_followers, author_gender, author_location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6fc942b9-dc15-459d-8e2b-13a55404a649",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_raw_text(weibo):\n",
    "    iffull = weibo.xpath(\".//a[@action-type='fl_unfold']\")\n",
    "    if iffull:\n",
    "        text_raw = weibo.xpath(\".//p[@node-type='feed_list_content_full']\")[0] # 'feed_list_content_full' for imcompleted weibos\n",
    "    else:\n",
    "        text_raw = weibo.xpath(\".//p[@node-type='feed_list_content']\")[0]\n",
    "    return text_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "32d118ff-703a-43c8-87a5-c5eb17d28d46",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_time(weibo):\n",
    "    try:\n",
    "        time = get_text(weibo.xpath(\".//div[@class='from']/a\")[0].xpath(\"string()\")).translate(mytable)\n",
    "    except:\n",
    "        time = get_text(weibo.xpath(\".//p[@class='from']/a\")[0].xpath(\"string()\")).translate(mytable)\n",
    "        \n",
    "    if time[1].isdigit():\n",
    "        if int(time[0:2]) < 20:\n",
    "            time = '2023-'+time\n",
    "    else:\n",
    "        time = '2023-05'\n",
    "    return time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dc62d4cc-e061-4f75-9f67-d1b6954d5d7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_interaction(weibo):\n",
    "    if len(weibo.xpath(\".//a[@action-type='feed_list_like']/em\")) == 0:\n",
    "        like = weibo.xpath(\".//div[@class='card-act']//span[@class='woo-like-count']\")[0].text.replace('赞','')\n",
    "        comment = get_text(weibo.xpath(\".//a[@action-type='feed_list_comment']\")[0].xpath(\"string()\").replace('评论',''))\n",
    "        repost = get_text(weibo.xpath(\".//a[@action-type='feed_list_forward']\")[0].xpath(\"string()\").replace('转发',''))\n",
    "    else: # old version\n",
    "        like = weibo.xpath(\".//a[@action-type='feed_list_like']/em\")[0].text\n",
    "        comment = get_text(weibo.xpath(\".//a[@action-type='feed_list_comment']\")[0].text).replace('评论','')\n",
    "        repost = get_text(weibo.xpath(\".//a[@action-type='feed_list_forward']\")[0].text).replace('转发','')\n",
    "    return like,comment,repost"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a1fef2ee-bc0b-4e69-b844-ddf40dd2c17d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Update cookie\n",
    "with open('weibo/cookie.txt', 'r') as f:\n",
    "    Cookie = f.readlines()[0]  # Read all lines in the file\n",
    "User_Agent = \"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.3 Safari/605.1.15\"\n",
    "headers = {\n",
    "            'Cookie': Cookie,\n",
    "            'User-Agent': User_Agent\n",
    "        }\n",
    "\n",
    "savefile = 'weibo/weibo.csv'\n",
    "infofile = 'weibo/info.txt'\n",
    "if os.path.exists(infofile):\n",
    "    with open(infofile,'r') as f:\n",
    "        s = f.read()\n",
    "        my_dict = json.loads(s)\n",
    "start_time = my_dict['start_time']\n",
    "end_time = my_dict['end_time']\n",
    "keyword = my_dict['keyword']\n",
    "page = my_dict['page']\n",
    "mytable = start_time.maketrans('年月日:','----')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f40e5ad6-a430-4aab-a9d2-b65d8b3605b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-24-15&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-24-14&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-24-13&Refer=g&page=25\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/8v/xpk1jpgx2kbgxh9xl6dwblt00000gp/T/ipykernel_85649/3202781325.py\", line 41, in <module>\n",
      "    author_following, author_followers, author_gender, author_location = get_author_info(uid)\n",
      "  File \"/var/folders/8v/xpk1jpgx2kbgxh9xl6dwblt00000gp/T/ipykernel_85649/1948625335.py\", line 6, in get_author_info\n",
      "    info_dict = json.loads(res_author)['data']['user']\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-24-12&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-24-11&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-24-10&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-24-09&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-24-08&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-24-06&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-24-02&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-24-00&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-23&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-22&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-21&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-20&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-19&Refer=g&page=25\n",
      "time:  2023-05-17-23-08 . Please check the time\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-18&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-17&Refer=g&page=25\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 699, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 382, in _make_request\n",
      "    self._validate_conn(conn)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 1010, in _validate_conn\n",
      "    conn.connect()\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connection.py\", line 416, in connect\n",
      "    self.sock = ssl_wrap_socket(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/util/ssl_.py\", line 449, in ssl_wrap_socket\n",
      "    ssl_sock = _ssl_wrap_socket_impl(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/util/ssl_.py\", line 493, in _ssl_wrap_socket_impl\n",
      "    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/ssl.py\", line 500, in wrap_socket\n",
      "    return self.sslsocket_class._create(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/ssl.py\", line 1040, in _create\n",
      "    self.do_handshake()\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/ssl.py\", line 1309, in do_handshake\n",
      "    self._sslobj.do_handshake()\n",
      "ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:1129)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/adapters.py\", line 439, in send\n",
      "    resp = conn.urlopen(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 755, in urlopen\n",
      "    retries = retries.increment(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/util/retry.py\", line 574, in increment\n",
      "    raise MaxRetryError(_pool, url, error or ResponseError(cause))\n",
      "urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='weibo.com', port=443): Max retries exceeded with url: /ajax/profile/info?custom=6511100475 (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/8v/xpk1jpgx2kbgxh9xl6dwblt00000gp/T/ipykernel_85649/3202781325.py\", line 41, in <module>\n",
      "    author_following, author_followers, author_gender, author_location = get_author_info(uid)\n",
      "  File \"/var/folders/8v/xpk1jpgx2kbgxh9xl6dwblt00000gp/T/ipykernel_85649/1948625335.py\", line 4, in get_author_info\n",
      "    res_author = requests.get(url, headers=headers).text\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/api.py\", line 75, in get\n",
      "    return request('get', url, params=params, **kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/api.py\", line 61, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/sessions.py\", line 542, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/sessions.py\", line 655, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/adapters.py\", line 514, in send\n",
      "    raise SSLError(e, request=request)\n",
      "requests.exceptions.SSLError: HTTPSConnectionPool(host='weibo.com', port=443): Max retries exceeded with url: /ajax/profile/info?custom=6511100475 (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n",
      "\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-16&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-15&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-14&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-13&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-12&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-11&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-10&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-09&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-08&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-06&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-02&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-23-00&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-23&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-22&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-21&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-20&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-19&Refer=g&page=25\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 699, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 445, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 440, in _make_request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/http/client.py\", line 1371, in getresponse\n",
      "    response.begin()\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/http/client.py\", line 319, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/http/client.py\", line 288, in _read_status\n",
      "    raise RemoteDisconnected(\"Remote end closed connection without\"\n",
      "http.client.RemoteDisconnected: Remote end closed connection without response\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/adapters.py\", line 439, in send\n",
      "    resp = conn.urlopen(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 755, in urlopen\n",
      "    retries = retries.increment(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/util/retry.py\", line 532, in increment\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/packages/six.py\", line 769, in reraise\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 699, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 445, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 440, in _make_request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/http/client.py\", line 1371, in getresponse\n",
      "    response.begin()\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/http/client.py\", line 319, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/http/client.py\", line 288, in _read_status\n",
      "    raise RemoteDisconnected(\"Remote end closed connection without\"\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/8v/xpk1jpgx2kbgxh9xl6dwblt00000gp/T/ipykernel_85649/3202781325.py\", line 41, in <module>\n",
      "    author_following, author_followers, author_gender, author_location = get_author_info(uid)\n",
      "  File \"/var/folders/8v/xpk1jpgx2kbgxh9xl6dwblt00000gp/T/ipykernel_85649/1948625335.py\", line 4, in get_author_info\n",
      "    res_author = requests.get(url, headers=headers).text\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/api.py\", line 75, in get\n",
      "    return request('get', url, params=params, **kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/api.py\", line 61, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/sessions.py\", line 542, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/sessions.py\", line 655, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/adapters.py\", line 498, in send\n",
      "    raise ConnectionError(err, request=request)\n",
      "requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/8v/xpk1jpgx2kbgxh9xl6dwblt00000gp/T/ipykernel_85649/3202781325.py\", line 41, in <module>\n",
      "    author_following, author_followers, author_gender, author_location = get_author_info(uid)\n",
      "  File \"/var/folders/8v/xpk1jpgx2kbgxh9xl6dwblt00000gp/T/ipykernel_85649/1948625335.py\", line 6, in get_author_info\n",
      "    info_dict = json.loads(res_author)['data']['user']\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-18&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-17&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-16&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-15&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-14&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-13&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-12&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-11&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-10&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-09&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-07&Refer=g&page=25\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 699, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 382, in _make_request\n",
      "    self._validate_conn(conn)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 1010, in _validate_conn\n",
      "    conn.connect()\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connection.py\", line 416, in connect\n",
      "    self.sock = ssl_wrap_socket(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/util/ssl_.py\", line 449, in ssl_wrap_socket\n",
      "    ssl_sock = _ssl_wrap_socket_impl(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/util/ssl_.py\", line 493, in _ssl_wrap_socket_impl\n",
      "    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/ssl.py\", line 500, in wrap_socket\n",
      "    return self.sslsocket_class._create(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/ssl.py\", line 1040, in _create\n",
      "    self.do_handshake()\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/ssl.py\", line 1309, in do_handshake\n",
      "    self._sslobj.do_handshake()\n",
      "ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:1129)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/adapters.py\", line 439, in send\n",
      "    resp = conn.urlopen(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 755, in urlopen\n",
      "    retries = retries.increment(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/util/retry.py\", line 574, in increment\n",
      "    raise MaxRetryError(_pool, url, error or ResponseError(cause))\n",
      "urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='weibo.com', port=443): Max retries exceeded with url: /ajax/profile/info?custom=1747695872 (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/8v/xpk1jpgx2kbgxh9xl6dwblt00000gp/T/ipykernel_85649/3202781325.py\", line 41, in <module>\n",
      "    author_following, author_followers, author_gender, author_location = get_author_info(uid)\n",
      "  File \"/var/folders/8v/xpk1jpgx2kbgxh9xl6dwblt00000gp/T/ipykernel_85649/1948625335.py\", line 4, in get_author_info\n",
      "    res_author = requests.get(url, headers=headers).text\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/api.py\", line 75, in get\n",
      "    return request('get', url, params=params, **kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/api.py\", line 61, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/sessions.py\", line 542, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/sessions.py\", line 655, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/adapters.py\", line 514, in send\n",
      "    raise SSLError(e, request=request)\n",
      "requests.exceptions.SSLError: HTTPSConnectionPool(host='weibo.com', port=443): Max retries exceeded with url: /ajax/profile/info?custom=1747695872 (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n",
      "\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-02&Refer=g&page=25\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 699, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 382, in _make_request\n",
      "    self._validate_conn(conn)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 1010, in _validate_conn\n",
      "    conn.connect()\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connection.py\", line 416, in connect\n",
      "    self.sock = ssl_wrap_socket(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/util/ssl_.py\", line 449, in ssl_wrap_socket\n",
      "    ssl_sock = _ssl_wrap_socket_impl(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/util/ssl_.py\", line 493, in _ssl_wrap_socket_impl\n",
      "    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/ssl.py\", line 500, in wrap_socket\n",
      "    return self.sslsocket_class._create(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/ssl.py\", line 1040, in _create\n",
      "    self.do_handshake()\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/ssl.py\", line 1309, in do_handshake\n",
      "    self._sslobj.do_handshake()\n",
      "ConnectionResetError: [Errno 54] Connection reset by peer\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/adapters.py\", line 439, in send\n",
      "    resp = conn.urlopen(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 755, in urlopen\n",
      "    retries = retries.increment(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/util/retry.py\", line 532, in increment\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/packages/six.py\", line 769, in reraise\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 699, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 382, in _make_request\n",
      "    self._validate_conn(conn)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 1010, in _validate_conn\n",
      "    conn.connect()\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/connection.py\", line 416, in connect\n",
      "    self.sock = ssl_wrap_socket(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/util/ssl_.py\", line 449, in ssl_wrap_socket\n",
      "    ssl_sock = _ssl_wrap_socket_impl(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/urllib3/util/ssl_.py\", line 493, in _ssl_wrap_socket_impl\n",
      "    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/ssl.py\", line 500, in wrap_socket\n",
      "    return self.sslsocket_class._create(\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/ssl.py\", line 1040, in _create\n",
      "    self.do_handshake()\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/ssl.py\", line 1309, in do_handshake\n",
      "    self._sslobj.do_handshake()\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(54, 'Connection reset by peer'))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/8v/xpk1jpgx2kbgxh9xl6dwblt00000gp/T/ipykernel_85649/3202781325.py\", line 41, in <module>\n",
      "    author_following, author_followers, author_gender, author_location = get_author_info(uid)\n",
      "  File \"/var/folders/8v/xpk1jpgx2kbgxh9xl6dwblt00000gp/T/ipykernel_85649/1948625335.py\", line 4, in get_author_info\n",
      "    res_author = requests.get(url, headers=headers).text\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/api.py\", line 75, in get\n",
      "    return request('get', url, params=params, **kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/api.py\", line 61, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/sessions.py\", line 542, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/sessions.py\", line 655, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"/Users/liaoyufang/opt/anaconda3/lib/python3.9/site-packages/requests/adapters.py\", line 498, in send\n",
      "    raise ConnectionError(err, request=request)\n",
      "requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(54, 'Connection reset by peer'))\n",
      "\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-22-00&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-23&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-22&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-21&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-20&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-19&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-18&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-17&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-16&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-15&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-14&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-13&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-12&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-11&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-10&Refer=g&page=25\n",
      "https://s.weibo.com/weibo?q=%E5%A5%B3%E6%9D%83&typeall=1&suball=1&timescope=custom%3A2021-02-01-00%3A2021-03-21-08&Refer=g&page=25\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/8v/xpk1jpgx2kbgxh9xl6dwblt00000gp/T/ipykernel_85649/3202781325.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     11\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpage\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0;36m25\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mcontinue_scrape\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m             \u001b[0murl\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'https://s.weibo.com/weibo?q={}&typeall=1&suball=1&timescope=custom%3A{}%3A{}&Refer=g&page={}'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mquote\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeyword\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mstart_time\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mend_time\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpage\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m             \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrequests\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheaders\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mheaders\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m             \u001b[0mhtml\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0metree\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mHTML\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mres\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtext\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'utf-8'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/requests/api.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     73\u001b[0m     \"\"\"\n\u001b[1;32m     74\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 75\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'get'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     76\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     77\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/requests/api.py\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     59\u001b[0m     \u001b[0;31m# cases, and look like a memory leak in others.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     60\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0msessions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSession\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0msession\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 61\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0msession\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     62\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/requests/sessions.py\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    540\u001b[0m         }\n\u001b[1;32m    541\u001b[0m         \u001b[0msend_kwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msettings\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 542\u001b[0;31m         \u001b[0mresp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprep\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0msend_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    543\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    544\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/requests/sessions.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    653\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    654\u001b[0m         \u001b[0;31m# Send the request\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 655\u001b[0;31m         \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madapter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    656\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    657\u001b[0m         \u001b[0;31m# Total elapsed time of the request (approximately)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/requests/adapters.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    437\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    438\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mchunked\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 439\u001b[0;31m                 resp = conn.urlopen(\n\u001b[0m\u001b[1;32m    440\u001b[0m                     \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    441\u001b[0m                     \u001b[0murl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    697\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    698\u001b[0m             \u001b[0;31m# Make the request on the httplib connection object.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 699\u001b[0;31m             httplib_response = self._make_request(\n\u001b[0m\u001b[1;32m    700\u001b[0m                 \u001b[0mconn\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    701\u001b[0m                 \u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36m_make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    380\u001b[0m         \u001b[0;31m# Trigger any extra validation we need to do.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    381\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 382\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_conn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    383\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mSocketTimeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mBaseSSLError\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    384\u001b[0m             \u001b[0;31m# Py2 raises this as a BaseSSLError, Py3 raises it as socket timeout.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36m_validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m   1008\u001b[0m         \u001b[0;31m# Force connect early to allow us to validate the connection.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1009\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"sock\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# AppEngine might not have  `.sock`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1010\u001b[0;31m             \u001b[0mconn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconnect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1011\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1012\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mconn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_verified\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/urllib3/connection.py\u001b[0m in \u001b[0;36mconnect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    414\u001b[0m             \u001b[0mcontext\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload_default_certs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    415\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 416\u001b[0;31m         self.sock = ssl_wrap_socket(\n\u001b[0m\u001b[1;32m    417\u001b[0m             \u001b[0msock\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconn\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    418\u001b[0m             \u001b[0mkeyfile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkey_file\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/urllib3/util/ssl_.py\u001b[0m in \u001b[0;36mssl_wrap_socket\u001b[0;34m(sock, keyfile, certfile, cert_reqs, ca_certs, server_hostname, ssl_version, ciphers, ssl_context, ca_cert_dir, key_password, ca_cert_data, tls_in_tls)\u001b[0m\n\u001b[1;32m    447\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    448\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0msend_sni\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 449\u001b[0;31m         ssl_sock = _ssl_wrap_socket_impl(\n\u001b[0m\u001b[1;32m    450\u001b[0m             \u001b[0msock\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcontext\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtls_in_tls\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mserver_hostname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mserver_hostname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    451\u001b[0m         )\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/urllib3/util/ssl_.py\u001b[0m in \u001b[0;36m_ssl_wrap_socket_impl\u001b[0;34m(sock, ssl_context, tls_in_tls, server_hostname)\u001b[0m\n\u001b[1;32m    491\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    492\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mserver_hostname\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 493\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mssl_context\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrap_socket\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msock\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mserver_hostname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mserver_hostname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    494\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    495\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mssl_context\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrap_socket\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msock\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/ssl.py\u001b[0m in \u001b[0;36mwrap_socket\u001b[0;34m(self, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, session)\u001b[0m\n\u001b[1;32m    498\u001b[0m         \u001b[0;31m# SSLSocket class handles server_hostname encoding before it calls\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    499\u001b[0m         \u001b[0;31m# ctx._wrap_socket()\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 500\u001b[0;31m         return self.sslsocket_class._create(\n\u001b[0m\u001b[1;32m    501\u001b[0m             \u001b[0msock\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msock\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    502\u001b[0m             \u001b[0mserver_side\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mserver_side\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/ssl.py\u001b[0m in \u001b[0;36m_create\u001b[0;34m(cls, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, context, session)\u001b[0m\n\u001b[1;32m   1038\u001b[0m                         \u001b[0;31m# non-blocking\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1039\u001b[0m                         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"do_handshake_on_connect should not be specified for non-blocking sockets\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1040\u001b[0;31m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdo_handshake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1041\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mOSError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1042\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/ssl.py\u001b[0m in \u001b[0;36mdo_handshake\u001b[0;34m(self, block)\u001b[0m\n\u001b[1;32m   1307\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0.0\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mblock\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1308\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msettimeout\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1309\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sslobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdo_handshake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1310\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1311\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msettimeout\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "continue_scrape = 1\n",
    "\n",
    "with open(savefile, 'a+') as f:\n",
    "    header = ['mid','text','time','favorite','like','comment','repost','author_uid','author_url','author_following',\n",
    "              'author_followers','author_gender','author_location']\n",
    "    writer = csv.DictWriter(f, delimiter=',', lineterminator='\\n',fieldnames=header)\n",
    "    if not os.path.getsize(savefile):\n",
    "        writer.writeheader()\n",
    "\n",
    "    while(continue_scrape):\n",
    "        while ((page <= 25) & (continue_scrape)):\n",
    "            url = 'https://s.weibo.com/weibo?q={}&typeall=1&suball=1&timescope=custom%3A{}%3A{}&Refer=g&page={}'.format(quote(keyword),start_time,end_time,page)\n",
    "            res = requests.get(url=url, headers=headers)\n",
    "\n",
    "            html = etree.HTML(res.text.encode('utf-8'))\n",
    "\n",
    "            try:\n",
    "                weibos = html.xpath(\"//div[@action-type='feed_list_item']\")\n",
    "\n",
    "                # 自动结束\n",
    "                if len(weibos) == 0:\n",
    "#                    continue_scrape = 0\n",
    "                    print(f'no weibo, current time: {end_time}, page: {page}\\n')\n",
    "                    page += 1\n",
    "                    continue\n",
    "\n",
    "                for weibo in weibos:\n",
    "                    mid = weibo.xpath(\"./@mid\")[0]\n",
    "\n",
    "                    text_raw = get_raw_text(weibo)\n",
    "                    text = get_text(text_raw.xpath(\"string()\").replace('收起全文d',''))\n",
    "                    time = get_time(weibo)\n",
    "                    if time[0:4] == '2023':\n",
    "                        print('time: ', time, '. Please check the time')\n",
    "                        break\n",
    "                    favorite = ''\n",
    "                    like,comment,repost = get_interaction(weibo)\n",
    "\n",
    "                    uid = re.search(r'/(.*)\\?refer',weibo.xpath(\".//a/@href\")[0].replace('//',''), re.IGNORECASE).group(1)\n",
    "                    author_url = f'https://weibo.com/p/100505{uid}/info?mod=pedit_more'\n",
    "                    author_following, author_followers, author_gender, author_location = get_author_info(uid)\n",
    "#                    author_following, author_followers, author_gender, author_location = 0,0,'',''\n",
    "\n",
    "                    writer.writerow({'mid':mid,'text':text,'time':time,'favorite':favorite,'like':like,'comment':comment,'repost':repost,\n",
    "                                       'author_uid':uid,'author_url':author_url, 'author_following':author_following, 'author_followers':author_followers, \n",
    "                                        'author_gender':author_gender, 'author_location':author_location})\n",
    "            except:\n",
    "                print(traceback.format_exc())\n",
    "            page += 1\n",
    "            sleep(random.uniform(4, 6))\n",
    "            with open(infofile, 'w') as f:\n",
    "                f.write('{\"keyword\": \"女权\", \"start_time\": \"'+start_time+'\", '+'\"end_time\": \"{}\", \"page\":{}'.format(end_time,page)+'}')\n",
    "        print(url)\n",
    "        sleep(10)\n",
    "        if time[0:4] != '2023':\n",
    "            end_time = time[0:-3] #be careful about duplicates since its based on hour instead of minute\n",
    "        page = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f3c68df-1a78-4674-9b0c-47791c23799c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#old version\n",
    "    # if match_following := re.search(r'<strong class=\"W_f..\">(.*)</strong><span class=\"S_txt2\">关注',res_author, re.IGNORECASE):\n",
    "    #     author_following = match_following.group(1)\n",
    "    # if match_followers := re.search(r'关注.*<strong class=\"W_f..\">(.*)</strong><span class=\"S_txt2\">粉丝',res_author, re.IGNORECASE):\n",
    "    #     author_followers = match_followers.group(1)\n",
    "    # if match_gender := re.search(r'性别：</span><span class=\"pt_detail\">(.)</span></li>',res_author, re.IGNORECASE):\n",
    "    #     author_gender = match_gender.group(1)\n",
    "    # if match_location := re.search(r'所在地：</span><span class=\"pt_detail\">(.*)</span>.*IP属地：',res_author, re.IGNORECASE):\n",
    "    #     author_location = match_location.group(1)\n",
    "    # if match_IPlocation := re.search(r'IP属地：</span><span class=\"pt_detail ipRegion\">(.*)<span class=\"W_icon icon_askS\">',res_author, re.IGNORECASE):\n",
    "    #     author_IPlocation = match_IPlocation.group(1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
